src/read.c

```diff
@@ -211,185 +211,452 @@ static Obj GAPInfo;
+**
+**  R_LVAR:             local var with id <var>
+**  R_HVAR:             high var with id <var>
+**  R_DVAR:             debug var with id <var>, at nesting level <nest0>
+**  R_GVAR:             global var with id <var>
+**  R_ELM_LIST:         uses <narg>, <level>
+**  R_ELMS_LIST:        uses <level>
+**  R_ELM_POSOBJ:       uses <level>
+**  R_ELMS_POSOBJ:      uses <level>
+**  R_ELM_REC_NAME:     record access r.<rnam>
+**  R_ELM_REC_EXPR      record access r.(expr)
+**  R_ELM_COMOBJ_NAME:  com obj access obj.<rnam>
+**  R_ELM_COMOBJ_EXPR:  com obj access obj.(expr)
+**  R_FUNCCALL          function call without options and with <narg> arguments
+**  R_FUNCCALL_OPTS     function call with options and with <narg> arguments
+*/
+enum REFTYPE {
+    R_INVALID,
+    R_LVAR,
+    R_HVAR,
+    R_DVAR,
+    R_GVAR,
+    R_ELM_LIST,
+    R_ELMS_LIST,
+    R_ELM_POSOBJ,
+    R_ELMS_POSOBJ,
+    R_ELM_REC_NAME,
+    R_ELM_REC_EXPR,
+    R_ELM_COMOBJ_NAME,
+    R_ELM_COMOBJ_EXPR,
+    R_FUNCCALL,
+    R_FUNCCALL_OPTS,
+};
+
+typedef struct {
+    enum REFTYPE type;
+
+    union {
+        UInt var;
+        UInt narg;
+        UInt rnam;
+    };
 
-void ReadReferenceModifiers( TypSymbolSet follow )
+    union {
+        UInt nest0;
+        UInt level;
+    };
+} LHSRef;
+
+
+/****************************************************************************
+**
+*/
+UInt EvalRef(const LHSRef ref, Int needExpr)
 {
-    volatile char type = ' ';
-    volatile UInt level = 0;
-    volatile UInt narg = 0;
-    volatile UInt rnam  = 0;
-    /* followed by one or more selectors                                   */
-    while ( IS_IN( STATE(Symbol), S_LPAREN|S_LBRACK|S_LBRACE|S_DOT ) ) {
-        /* <Var> '[' <Expr> ']'  list selector                             */
-        if ( STATE(Symbol) == S_LBRACK ) {
-            Match( S_LBRACK, "[", follow );
-            ReadExpr( S_COMMA|S_RBRACK|follow, 'r' );
-            narg = 1;
-            while ( STATE(Symbol) == S_COMMA) {
-              Match(S_COMMA,",", follow|S_RBRACK);
-              ReadExpr(S_COMMA|S_RBRACK|follow, 'r' );
-              narg++;
-            }
-            Match( S_RBRACK, "]", follow );
-            type = (level == 0 ? '[' : ']');
+    TRY_READ
+    {
+        switch (ref.type) {
+        case R_LVAR:
+            IntrRefLVar(ref.var);
+            break;
+        case R_HVAR:
+            IntrRefHVar(ref.var);
+            break;
+        case R_DVAR:
+            IntrRefDVar(ref.var, ref.nest0);
+            break;
+        case R_GVAR:
+            IntrRefGVar(ref.var);
+            break;
+        case R_ELM_LIST:
+            if (ref.level == 0)
+                IntrElmList(ref.narg);
+            else
+                IntrElmListLevel(ref.narg, ref.level);
+            break;
+        case R_ELMS_LIST:
+            if (ref.level == 0)
+                IntrElmsList();
+            else
+                IntrElmsListLevel(ref.level);
+            return ref.level + 1;
```

Comment:
Note that here is a slight behavior change, as we are inside a `TRY_READ` block: in the old code, if `level` was >0 and `IntrElmsListLevel` triggered an error, we simply did not increment `level`, and hence ended up using the same `level` value as before. Now, we get to the `return 0` before, i.e. effectively reset the level to 0.

I don't know if this could cause any problems, I have not thought too much about examples where this might cause a problem. If somebody knows of some examples, please share them!